### ISSUE TITLE
System.Net.Sockets tests crash (exception) about 1 in 2 runs on my dev machine

### ISSUE THREAD

**Issue by CIPop** [dotnet/corefx#19428]
```
0:041> !pe -nested
Exception object: 000001cf8048d7c8
Exception type:   System.Net.Sockets.SocketException
Message:          An existing connection was forcibly closed by the remote host
InnerException:   <none>
StackTrace (generated):
    SP               IP               Function
    0000003A8977F280 00007FFD130A5E3A System_Net_Sockets!System.Net.Sockets.Socket.Receive(Byte[], Int32, Int32, System.Net.Sockets.SocketFlags)+0xaa
    0000003A8977F2F0 00007FFD1315352A System_Net_Sockets_Tests!System.Net.Sockets.Tests.SendFileTest+<>c__DisplayClass5_0.<SendFile_Synchronous>b__0()+0xba
    0000003A8977F380 00007FFD131528AC System_Threading_Thread!System.Threading.Thread.ThreadMain_ThreadStart()+0xfc
    0000003A8977F3F0 00007FFD720DEFA1 System_Private_CoreLib!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)+0x111

StackTraceString: <none>
HResult: 80004005

Nested exception -------------------------------------------------------------
Exception object: 000001cf8048d7c8
Exception type:   System.Net.Sockets.SocketException
Message:          An existing connection was forcibly closed by the remote host
InnerException:   <none>
StackTrace (generated):
    SP               IP               Function
    0000003A8977F280 00007FFD130A5E3A System_Net_Sockets!System.Net.Sockets.Socket.Receive(Byte[], Int32, Int32, System.Net.Sockets.SocketFlags)+0xaa
    0000003A8977F2F0 00007FFD1315352A System_Net_Sockets_Tests!System.Net.Sockets.Tests.SendFileTest+<>c__DisplayClass5_0.<SendFile_Synchronous>b__0()+0xba
    0000003A8977F380 00007FFD131528AC System_Threading_Thread!System.Threading.Thread.ThreadMain_ThreadStart()+0xfc
    0000003A8977F3F0 00007FFD720DEFA1 System_Private_CoreLib!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)+0x111

StackTraceString: <none>
HResult: 80004005
```

**Comment by CIPop** [comment]
Full-memory dump available (too big, can't attach).

**Comment by stephentoub** [comment]
The client is disposing of the Socket here:
https://github.com/dotnet/corefx/blob/654ad09c9313bffef923351216801ef626f55d6a/src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L157
which is apparently causing the Receive here:
https://github.com/dotnet/corefx/blob/654ad09c9313bffef923351216801ef626f55d6a/src/System.Net.Sockets/tests/FunctionalTests/SendFile.cs#L133
to fail with that error.  Does the client need to explicitly Shutdown prior to the implicit Dispose as part of the using?

**Comment by CIPop** [comment]
> Does the client need to explicitly Shutdown prior to the implicit Dispose as part of the using?

Thanks for looking into this @stephentoub ! 

You're right: the deterministic way to achieve graceful termination requires all parties to `shutdown` at least their side causing the other side's `read` to return 0 (EOF). 
I'm not 100% sure but I think the time set here is likely the cause: `client.LingerState = new LingerOption(true, LingerTime);` which will wait 10 seconds (`LingerTime`) until Close is called. 

/cc @schen381

**Comment by schen381** [comment]
Looks similar to this issue here: https://github.com/dotnet/corefx/issues/13114#issuecomment-290554256. I will try to repro on my machine.

**Comment by schen381** [comment]
After investigating into #19428, there is a risk by using LingerState to wait for send/receive operation complete, before dispose the socket.

Because the LingerTime value is arbitrarily set by us, some slow computers may take more time to finish the send/receive operation, so it's possible that some data are still available before we dispose the socket. And in this case, we will use a RST packet (instead of FIN) to forcibly close the connection after time out (https://github.com/dotnet/corefx/blob/master/src/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4649, related issue #13114), and test will fail with System.Net.Sockets.SocketException, Message: An existing connection was forcibly closed by the remote host. An example is #19428

Potential fix: always use socket.Shutdown(SocketShutdown how) before we dispose() the socket.

**Comment by karelz** [comment]
@schen381 only one issue type please - either "test bug" or "test enhancement" -- see [triage rules](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#triage-rules---simplified)

**Comment by schen381** [comment]
@karelz Sorry, since I opened a broader discussion for this issue in the comment, I am not sure how to triage it. @davidsh could you take a look please?

**Comment by davidsh** [comment]
We should stay with "test bug" for now since we still unreliable tests at times due to the currently test implementation. I've updated the labels.

**Comment by davidsh** [comment]
This no longer repros.  I believe it was fixed with #19603 which added using `socket.Shutdown()` to provide a clean dispose and FIN of the socket.
